File: Leetcode - Python/554_map_leastBricks.py
# https://leetcode.com/problems/brick-wall/

import sys
import collections


class Solution:
    def leastBricks(self, wall: List[List[int]]) -> int:
        res = -sys.maxsize - 1
        m = collections.defaultdict(int)

        for row in wall:
            width = 0
            for b in row[: -1]:
                width += b
                m[width] += 1

        return len(wall) - max(m.values(), default=0)

File: Leetcode - Java/554_map_leastBricks.java
// https://leetcode.com/problems/brick-wall/

class Solution {
    public int leastBricks(List<List<Integer>> wall) {
        int res = Integer.MIN_VALUE;
        Map<Integer, Integer> map = new HashMap();
        
        for (List<Integer> w : wall) {
            int width = 0;
            for (int i = 0; i < w.size() - 1; i++) {
                width += w.get(i);
                map.put(width, map.getOrDefault(width, 0) + 1);
            }
        }
        
        for (Map.Entry<Integer, Integer> m : map.entrySet()) {
            res = Math.max(res, m.getValue());
        }
        
        if (res == Integer.MIN_VALUE) res = 0;
        return wall.size() - res;
    }
}